src/check_presence.c

```diff
@@ -0,0 +1,533 @@
+      mdb_txn_abort(txn);
+      return 0;
+    }
+  }
+
+  mdb_dbi_close(env, dbi);
+  mdb_txn_abort(txn);
+
+  /*
+   * Slow path - we need to update the host's record.
+   *
+   * At this point the host is either brand new, or we have not seen this host
+   * during this second, and we need to perform a write to update the key
+   * value store.
+   *
+   * The LMDB key value store handles locking for us so that writes from
+   * multiple threads are serialised correctly. As soon as the write is
+   * complete, the subsequent reads follow the fast path above, keeping this
+   * as inexpensive as possible.
+   */
+
+  rc = mdb_txn_begin(env, NULL, 0, &txn);
+  if (rc) {
+    ERROR(PLUGIN_NAME " plugin: mdb_txn_begin returned: %s (%d)",
+          mdb_strerror(rc), rc);
+    return -1;
+  }
+
+  rc = mdb_dbi_open(txn, NULL, MDB_CREATE, &dbi);
+  if (rc) {
+    ERROR(PLUGIN_NAME " plugin: mdb_dbi_open returned: %s (%d)",
+          mdb_strerror(rc), rc);
+    mdb_txn_abort(txn);
+    return -1;
+  }
+
+  data.mv_size = sizeof(tv_now);
+  data.mv_data = &tv_now;
+
+  rc = mdb_put(txn, dbi, &key, &data, add ? MDB_NOOVERWRITE : 0);
+  if (MDB_KEYEXIST == rc) {
+
+    /* another thread beat us to it, do nothing */
+    add = 0;
+
+  } else if (rc) {
+    ERROR(PLUGIN_NAME " plugin: mdb_put returned: %s (%d)", mdb_strerror(rc),
+          rc);
+    mdb_dbi_close(env, dbi);
+    mdb_txn_abort(txn);
+    return -1;
+  }
+
+  rc = mdb_txn_commit(txn);
+  if (rc) {
+    ERROR(PLUGIN_NAME " plugin: mdb_txn_commit returned: %s (%d)",
+          mdb_strerror(rc), rc);
+    mdb_dbi_close(env, dbi);
+    return -1;
+  }
+
+  mdb_dbi_close(env, dbi);
+
+  /* finally, handle the notification if needed */
+  if (add) {
+    notification_t n = {0};
+    char message[NOTIF_MAX_MSG_LEN];
+
+    ssnprintf(message, sizeof(message), "Host is found: %.*s",
+              (int)sizeof(vl->host), vl->host);
+    notification_init(&n, NOTIF_OKAY, message, vl->host, PLUGIN_NAME, NULL,
+                      "host", "found");
+    n.time = cdtime();
+
+    plugin_dispatch_notification(&n);
+  }
+
+  return 0;
+}
+
+static int presence_config(const char *key, const char *value) {
+  if (strcasecmp(key, "STATEDATASTORE") == 0) {
+    if (value != NULL && strcmp(value, "") != 0) {
+      state_datastore = strdup(value);
+    } else {
+      state_datastore = NULL;
+    }
+    return 0;
+  } else if (strcasecmp(key, "HOSTTIMEOUT") == 0) {
+    if (value != NULL) {
+      host_timeout = atoi(value);
+    }
+    return 0;
+  } else if (strcasecmp(key, "STARTUPDELAY") == 0) {
+    if (value != NULL) {
+      startup_delay = atoi(value);
+    }
+    return 0;
+  } else
+    return -1;
+} /* int presence_config */
+
+void module_register(void) {
+  plugin_register_init(PLUGIN_NAME, presence_init);
+  plugin_register_config(PLUGIN_NAME, presence_config, config_keys,
+                         config_keys_num);
+  plugin_register_write(PLUGIN_NAME, presence_write, NULL);
```

Comment:
I think that is possible to move `register_write()` (and possibly `register_shutdown()`) into init phase and do not register it if not configured.
This will not waste CPU time.

Reply:
True - good to see this is an option.

Done.